libyul/optimiser/CommonSwitchCasePrefixMover.cpp

```diff
@@ -0,0 +1,161 @@
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libyul/optimiser/CommonSwitchCasePrefixMover.h>
+
+#include <libyul/optimiser/ASTCopier.h>
+#include <libyul/optimiser/NameCollector.h>
+#include <libyul/optimiser/SyntacticalEquality.h>
+#include <libyul/AST.h>
+#include <libsolutil/CommonData.h>
+
+#include <range/v3/algorithm/all_of.hpp>
+#include <range/v3/algorithm/transform.hpp>
+#include <range/v3/iterator.hpp>
+#include <range/v3/view/drop.hpp>
+#include <range/v3/view/transform.hpp>
+#include <range/v3/view/zip.hpp>
+#include <range/v3/action/transform.hpp>
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::yul;
+
+namespace
+{
+class IdentifierReplacer: public ASTCopier
+{
+public:
+	IdentifierReplacer(map<YulString, YulString> _identifierMap): m_identifierMap(move(_identifierMap)) {}
+	using ASTCopier::operator();
+protected:
+	YulString translateIdentifier(YulString _name) override
+	{
+		if (YulString const* value = util::valueOrNullptr(m_identifierMap, _name))
+			return *value;
+		else
+			return _name;
+	}
+private:
+	map<YulString, YulString> m_identifierMap;
+};
+}
+
+void CommonSwitchCasePrefixMover::run(OptimiserStepContext&, Block& _ast)
+{
+	CommonSwitchCasePrefixMover{}(_ast);
+}
+
+namespace
+{
+
+template<typename Container>
+auto eraseFirst(Container& _container) { return _container.erase(begin(_container)); }
+
+optional<Statement> tryExtractFirstStatement(Identifier const* _identifier, vector<Case>& _cases)
+{
+	Statement& referenceStatement = _cases.front().body.statements.front();
+
+	if (_identifier)
+	{
+		Assignments assignments;
+		visit(assignments, referenceStatement);
+		if (assignments.names().count(_identifier->name))
+			return nullopt;
+	}
+
+	if (!ranges::all_of(
+		_cases | ranges::cpp20::views::drop(1),
+		[&](Case& _case) {
+			return !_case.body.statements.empty() &&
+				SyntacticallyEqual{}(referenceStatement, _case.body.statements.front());
+		}
+	))
+		return nullopt;
+
+	if (VariableDeclaration const* referenceVarDecl = get_if<VariableDeclaration>(&referenceStatement))
+	{
+		for (Case& switchCase: _cases | ranges::cpp20::views::drop(1))
+		{
+			VariableDeclaration* varDecl = std::get_if<VariableDeclaration>(&switchCase.body.statements.front());
+			yulAssert(varDecl, "");
+			yulAssert(varDecl->variables.size() == referenceVarDecl->variables.size(), "");
+
+			static constexpr auto nameFromTypedName = [](TypedName const& _name) { return _name.name; };
+			auto range = ranges::views::zip(
+				varDecl->variables | ranges::views::transform(nameFromTypedName),
+				referenceVarDecl->variables | ranges::views::transform(nameFromTypedName)
+			);
```

Comment:
Could you please explain what this statement does? Googling didn't help much :( but I didn't spend too much time looking around :)

Reply:
``varDecl->variables | ranges::views::transform(nameFromTypedName),`` as well as ``referenceVarDecl->variables | ranges::views::transform(nameFromTypedName)`` are "range views" (i.e. abstract helpers allowing to use range-based-loops on them or to walk through them with iterators). Those ranges have the same size, in which case ``zip`` will just combine them to a "range view" of ``std::pair``s - which is exactly what's required to construct ``std::map``s. I can try to rewrite this a bit more readably, though.